core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java

```diff
@@ -0,0 +1,690 @@
+   * @param segmentFile
+   * @param path
+   * @throws IOException
+   */
+  public void writeSegmentFile(SegmentFile segmentFile, String path) throws IOException {
+    AtomicFileOperations fileWrite =
+        new AtomicFileOperationsImpl(path, FileFactory.getFileType(path));
+    BufferedWriter brWriter = null;
+    DataOutputStream dataOutputStream = null;
+    Gson gsonObjectToWrite = new Gson();
+    try {
+      dataOutputStream = fileWrite.openForWrite(FileWriteOperation.OVERWRITE);
+      brWriter = new BufferedWriter(new OutputStreamWriter(dataOutputStream,
+          Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET)));
+
+      String metadataInstance = gsonObjectToWrite.toJson(segmentFile);
+      brWriter.write(metadataInstance);
+    } finally {
+      if (null != brWriter) {
+        brWriter.flush();
+      }
+      CarbonUtil.closeStreams(brWriter);
+      fileWrite.close();
+    }
+  }
+
+  /**
+   * Merge all segment files in a segment to single file.
+   *
+   * @param writePath
+   * @throws IOException
+   */
+  public SegmentFile mergeSegmentFiles(String readPath, String mergeFileName, String writePath)
+      throws IOException {
+    CarbonFile[] segmentFiles = getSegmentFiles(readPath);
+    if (segmentFiles != null && segmentFiles.length > 0) {
+      SegmentFile segmentFile = null;
+      for (CarbonFile file : segmentFiles) {
+        SegmentFile localSegmentFile = readSegmentFile(file.getAbsolutePath());
+        if (segmentFile == null && localSegmentFile != null) {
+          segmentFile = localSegmentFile;
+        }
+        if (localSegmentFile != null) {
+          segmentFile = segmentFile.merge(localSegmentFile);
+        }
+      }
+      if (segmentFile != null) {
+        String path = writePath + "/" + mergeFileName + CarbonTablePath.SEGMENT_EXT;
+        writeSegmentFile(segmentFile, path);
+        FileFactory.deleteAllCarbonFilesOfDir(FileFactory.getCarbonFile(readPath));
+      }
+      return segmentFile;
+    }
+    return null;
+  }
+
+  private CarbonFile[] getSegmentFiles(String segmentPath) {
+    CarbonFile carbonFile = FileFactory.getCarbonFile(segmentPath);
+    if (carbonFile.exists()) {
+      return carbonFile.listFiles(new CarbonFileFilter() {
+        @Override public boolean accept(CarbonFile file) {
+          return file.getName().endsWith(CarbonTablePath.SEGMENT_EXT);
+        }
+      });
+    }
+    return null;
+  }
+
+  /**
+   * It provides segment file only for the partitions which has physical index files.
+   *
+   * @param partitionSpecs
+   */
+  public static SegmentFile getSegmentFileForPhysicalDataPartitions(String tablePath,
+      List<PartitionSpec> partitionSpecs) {
+    SegmentFile segmentFile = null;
+    for (PartitionSpec spec : partitionSpecs) {
+      String location = spec.getLocation().toString();
+      CarbonFile carbonFile = FileFactory.getCarbonFile(location);
+      boolean isRelative = false;
+      if (location.startsWith(tablePath)) {
```

Comment:
Move this if check after the below if check for {if (listFiles != null && listFiles.length > 0)}

Reply:
ok